### PR TITLE
Fix accountancy in block and vector builders

### PR DESCRIPTION
**Pull request by ChrisHegarty**
This commit fixes accountancy in block and vector builders.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st [hunk 1]
```diff
@@ -281,26 +281,17 @@ $if(BytesRef)$
...
         }
-        /*
-         * Update the breaker with the actual bytes used.
-         * We pass false below even though we've used the bytes. That's weird,
-         * but if we break here we will throw away the used memory, letting
-         * it be deallocated. The exception will bubble up and the builder will
-         * still technically be open, meaning the calling code should close it
-         * which will return all used memory to the breaker.
-         */
-        blockFactory.adjustBreaker(block.ramBytesUsed() - estimatedBytes, false);
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st:303]: Shouldn't we also remove the memory accounting from the ctor?

Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st:303]: I don't think so. The general idea here is to seed the initial memory adjustment based on the estimated size, then allow the memory to be adjusted incrementally as the builder has values added. The initial adjustment is done through `AbstractBlockBuilder::adjustBreaker`. Is it that the naked name `adjustBreaker` in the constructor is confusing? Or something else?

Comment (luigidellaquila) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st:303]: Uhm... yes, I see, you are also passing it to the factory as `preAdjustedBytes`. It's a bit confusing, but I think it's also fragile, eg. if the breaker fails, the builder remains open and the pre-allocated value is never released. 
Also, the factory has a problem with `adjustBreaker()`, see [this](https://github.com/elastic/elasticsearch/pull/100136), that makes the whole logic even more obscure IMHO

Comment (luigidellaquila) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st:303]: No, probably that's correct, but this mixed/shared manipulation of the circuit breaker is definitely confusing

**Review by nik9000 — APPROVED** [review]
If tests like it let's build on it. I like the idea of centralizing stuff for better mocking, including the breaker adjustment. For the fixed-length vector thingy we could even assert that the pre-adjusted bytes matches the built bytes. But that can wait for whenever.

**Pull request merged**